src/order/boolean_algebra.lean

```diff
@@ -107,31 +602,58 @@ theorem le_compl_iff_le_compl : y ≤ xᶜ ↔ x ≤ yᶜ :=
 theorem compl_le_iff_compl_le : xᶜ ≤ y ↔ yᶜ ≤ x :=
 ⟨compl_le_of_compl_le, compl_le_of_compl_le⟩
 
-theorem sup_sdiff_same : x ⊔ (y \ x) = x ⊔ y :=
-by simp [sdiff_eq, sup_inf_left]
-
-theorem sdiff_eq_left (h : x ⊓ y = ⊥) : x \ y = x :=
-by rwa [sdiff_eq, inf_eq_left, is_compl_compl.le_right_iff, disjoint_iff]
-
-theorem sdiff_le_sdiff (h₁ : w ≤ y) (h₂ : z ≤ x) : w \ x ≤ y \ z :=
-by rw [sdiff_eq, sdiff_eq]; from inf_le_inf h₁ (compl_le_compl h₂)
-
-@[simp] lemma sdiff_idem_right : x \ y \ y = x \ y :=
-by rw [sdiff_eq, sdiff_eq, inf_assoc, inf_idem]
-
 namespace boolean_algebra
 
 @[priority 100]
 instance : is_complemented α := ⟨λ x, ⟨xᶜ, is_compl_compl⟩⟩
 
 end boolean_algebra
 
+end boolean_algebra_core
+
+/-- A Boolean algebra is a bounded distributive lattice with
+a complement operator `ᶜ` such that `x ⊓ xᶜ = ⊥` and `x ⊔ xᶜ = ⊤`.
+For convenience, it must also provide a set difference operation `\`
+satisfying `x \ y = x ⊓ yᶜ`.
+
+This is a generalization of (classical) logic of propositions, or
+the powerset lattice. -/
+-- Lean complains about metavariables in the type if the universe is not specified
+class boolean_algebra (α : Type u) extends generalized_boolean_algebra α, boolean_algebra.core α :=
+(sdiff_eq : ∀x y:α, x \ y = x ⊓ yᶜ)
+-- TODO: is there a way to automatically fill in the proofs of sup_inf_sdiff and inf_inf_sdiff given
+-- everything in `boolean_algebra.core` and `sdiff_eq`? The following doesn't work:
+-- (sup_inf_sdiff := λ a b, by rw [sdiff_eq, ←inf_sup_left, sup_compl_eq_top, inf_top_eq])
```

Comment:
Is there a good way to get Lean to fill in proofs of these statements? Would we need a custom tactic of some sort?